hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+			}
+		}
+		recordStreamFileBuilder = null;
+		LOG.debug(OBJECT_STREAM.getMarker(), "RecordStreamFileWriter::clear executed.");
+	}
+
+	public void close() {
+		this.closeCurrentAndSign();
+		LOG.debug(LogMarker.FREEZE.getMarker(), "RecordStreamFileWriter finished writing the last object, is stopped");
+	}
+
+	public void setStartWriteAtCompleteWindow(final boolean startWriteAtCompleteWindow) {
+		this.startWriteAtCompleteWindow = startWriteAtCompleteWindow;
+		LOG.debug(OBJECT_STREAM.getMarker(),
+				"RecordStreamFileWriter::setStartWriteAtCompleteWindow: {}", startWriteAtCompleteWindow);
+	}
+
+	public boolean getStartWriteAtCompleteWindow() {
+		return this.startWriteAtCompleteWindow;
+	}
+
+	/**
+	 * Helper method that serializes a RecordStreamFile.
+	 * Uses deterministic serialization to ensure multiple invocations on the same
+	 * object lead to identical serialization.
+	 * @param recordStreamFileProtoBuilder
+	 * 		the RecordStreamFile object that needs to be serialized
+	 * @return
+	 * 		the serialized bytes
+	 */
+	private byte[] serialize(final RecordStreamFile.Builder recordStreamFileProtoBuilder) throws IOException {
+		final var recordStreamFileProto = recordStreamFileProtoBuilder.build();
+		final var result = new byte[recordStreamFileProto.getSerializedSize()];
+		final var output = CodedOutputStream.newInstance(result);
+		output.useDeterministicSerialization();
+		recordStreamFileProto.writeTo(output);
+		output.checkNoSpaceLeft();
+		return result;
+	}
+
+	private HashObject toProto(final Hash hash) {
+		return HashObject.newBuilder()
+				.setAlgorithm(HashAlgorithm.SHA_384)
+				.setLength(hash.getDigestType().digestLength())
+				.setHash(ByteString.copyFrom(hash.getValue()))
+				.build();
+	}
+
+	private void createSignatureFile(final File relatedRecordStreamFile) {
+		// get entire hash of current record stream file
+		final var entireHash = new Hash(streamDigest.digest(), DigestType.SHA_384);
+		// get metaData hash of current record stream file
+		final var metaHash = new Hash(metadataStreamDigest.digest(), DigestType.SHA_384);
+
+		// create proto messages for signature file
+		final var fileSignature = generateSignatureObject(entireHash);
+		final var metadataSignature = generateSignatureObject(metaHash);
+		final var signatureFile = SignatureFile.newBuilder()
+				.setFileSignature(fileSignature)
+				.setMetadataSignature(metadataSignature);
+
+		// create signature file
+		final var sigFilePath = generateSigFilePath(relatedRecordStreamFile);
+		try (final var fos = new FileOutputStream(sigFilePath)) {
+			fos.write(recordFileVersion); // version in signature files is 1 byte, compared to 4 in record files
+			signatureFile.build().writeTo(fos);
+			LOG.debug(OBJECT_STREAM_FILE.getMarker(),
+					"closeCurrentAndSign :: signature file saved: {}", sigFilePath);
+		} catch (IOException e) {
+			LOG.error(EXCEPTION.getMarker(),
```

Comment:
Is there any other action we can take when signature or record file writing fails? It's a pretty non-recoverable error so not sure an error log is sufficient.

Reply:
@tinker-michaelj wdyt?

Reply:
There are a few other places (e.g. [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/state/logic/BlockManager.java#L180)) where an exception is almost certainly irrecoverable.

In such places we currently do just log at `ERROR`; throwing an `AssertionError` could also be an option, we haven't set a policy really.

Reply:
Thanks Michael